src/models/train_model.py

```diff
@@ -0,0 +1,219 @@
+
+            vgg_feat = vgg_encoder(images)
+            mask = correct_capt == 0
+
+            # Generate Fake Images
+            fake_imgs, mu_tensor, logvar = generator(
+                noise,
+                sent_emb,
+                word_emb,
+                global_incept_feat,
+                local_incept_feat,
+                vgg_feat,
+                mask,
+            )
+
+            local_fake_incept_feat, global_fake_incept_feat = image_encoder(fake_imgs)
+            vgg_feat_fake = vgg_encoder(fake_imgs)
+
+            # Generate Logits for discriminator update
+            real_discri_feat = discriminator(images)
+            fake_discri_feat = discriminator(fake_imgs)
+
+            logits_discri = {
+                "fake": {
+                    "word_level": discriminator.get_word_level_logits(
+                        fake_discri_feat, word_emb
+                    ),
+                    "uncond": discriminator.get_uncond_logits(fake_discri_feat),
+                    "cond": discriminator.get_cond_logits(fake_discri_feat, sent_emb),
+                },
+                "real": {
+                    "word_level": discriminator.get_word_level_logits(
+                        real_discri_feat, word_emb
+                    ),
+                    "uncond": discriminator.get_uncond_logits(real_discri_feat),
+                    "cond": discriminator.get_cond_logits(real_discri_feat, sent_emb),
+                },
+            }
+
+            labels_discri = {
+                "fake": {"word_level": fake_word_labels, "image": labels_fake},
+                "real": {"word_level": word_labels, "image": labels_real},
+            }
+
+            # Update Discriminator
+            optimizer_d.zero_grad()
+            loss_discri = discriminator_loss(logits_discri, labels_discri, lambda4)
+
+            loss_discri.backward(retain_graph=True)
+            optimizer_d.step()
+
+            fake_feat_d = discriminator(fake_imgs)
+
+            logits_gen = {
+                "fake": {
+                    "word_level": discriminator.get_word_level_logits(
+                        fake_feat_d, word_emb
+                    ),
+                    "uncond": discriminator.get_uncond_logits(fake_feat_d),
+                    "cond": discriminator.get_cond_logits(fake_feat_d, sent_emb),
+                }
+            }
+
+            # Update Generator
+            optimizer_g.zero_grad()
+            loss_gen = generator_loss(
+                logits_gen,
+                local_fake_incept_feat,
+                global_fake_incept_feat,
+                labels_real,
+                word_labels,
+                word_emb,
+                sent_emb,
+                labels_match,
+                correct_capt_len,
+                curr_class,
+                vgg_feat,
+                vgg_feat_fake,
+                const_dict,
+            )
+
+            loss_kl = kl_loss(mu_tensor, logvar)
```

Comment:
why `kl_loss` is not a part of `generator_loss` function if it gets added together with `loss_gen` ?